PR #3 for User Context Implementation - EditingDomain adjustments

PhilMFischer opened the pull request:
This is the PR changing the _TransactionalEditingDomain_ related code for the multi user capabilities. Make sure that PR #571 and #576 with all data model related changes is merged first. 

This PR includes:
- Including the user context into transactional editing domain.
- Added method to CommandStack for finally executing a command in the user context.
- Fixed minor issues with taking the workspace lock usign the transactional editing domain (e.g. also in the DiagramHelper).
- Added test cases for the changes.
- Introduced new TestBuilder and Nature to detect invocations of the builder. Using the builders from virsat.build is not possible due to cyclic dependencies when executing with tycho/maven.
- Fixed some strange coverage issues on the DiagramHelper, likely a consequence of the changes there.


⚠️ _Atention ! Had to also work on the CodeCoverage of the graphiti plguin, i.e. the DiagramHelper._ ⚠️ 


This PR finally closes #548.

Commit:
Refactoring User Rights Management - (Task #548)

- Added new Interface for UserContext to UserRegistry
- Changed RightsHelper to make use of UserContext
- Added UserContext to TransactionalEditingDomain
- Adjusted TransactionalBuilders to make user of UserContext
- Adjusted Equation and InheritanceBuilder to make use of UserContext

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Fixed all code in test cases - (Task #548)

- Added new method to rights helper for checking writeAccess for
SIngleUser use cases

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Updated Model to use modified RMCheckCommand - (Task #548)

- Updated model.edit template
- run generation of model.edit
- Updated VirSatResourceSet to use Use Context
- Updated Transactional Editing Domain to use User Context
- Updated all relevant dependencies
- Fixed test cases

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Cleaning Up test cases - (Task #548)

- Cleaned Up Equation Builder test cases
- Added new test cases for VirSatEditingDomainRegistry

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Updated Assign Discipline Command - (Task #548)

- Refactored Code
- Improved test cases

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Added IUserContext to TED - (Task #548)

- Added user context to ted.
- Added override capabilities.
- Added mechanism to execute command in workspace of TED under different
user.
- Enhanced capabilities of Atomic Exception Throwing.

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Adjusting Command Stack for User Context - (Task #548)

- Refactored Command Stack to execute everything in one workspace
operation modified by a user context.

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Merge remote-tracking branch
'origin/feature/task_548-User_Context_Model_Adjustment' into
feature/task_548-User_Context_Implementation

Conflicts:
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/AAdvancedFunctionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/AExpressionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/ALeftOpRightExpressionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/ALiteralItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/AOpRightExpressionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/AdditionAndSubtractionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/AdvancedFunctionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/EquationDefinitionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/EquationIntermediateResultItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/EquationItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/EquationSectionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/FunctionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/IEquationDefinitionInputItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/IEquationDefinitionResultItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/IEquationInputItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/IEquationResultItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/ImportItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/MultiplicationAndDivisionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/NumberLiteralItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/ParenthesisItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/PowerFunctionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/ReferencedDefinitionInputItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/ReferencedInputItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/SetFunctionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/TypeDefinitionResultItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/TypeInstanceResultItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/ValueEItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/calculation/provider/ValuePiItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/APropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/AQudvTypePropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/BooleanPropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/ComposedPropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/DynamicArrayModifierItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/EReferencePropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/EnumPropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/EnumValueDefinitionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/FloatPropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/IntPropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/ReferencePropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/ResourcePropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/StaticArrayModifierItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertydefinitions/provider/StringPropertyItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/APropertyInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/ArrayInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/ComposedPropertyInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/EReferencePropertyInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/EnumUnitPropertyInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/ReferencePropertyInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/ResourcePropertyInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/UnitValuePropertyInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/propertyinstances/provider/ValuePropertyInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/provider/ATypeDefinitionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/provider/ATypeInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/provider/CategoryAssignmentItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/categories/provider/CategoryItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/concepts/provider/ConceptImportItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/concepts/provider/ConceptItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/concepts/provider/EcoreImportItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/concepts/provider/IEImportsItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/dmf/provider/DObjectContainerItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/dmf/provider/DObjectItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/dmf/provider/UnresolveableDObjectItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/provider/RepositoryItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/AConversionBasedUnitItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/AQuantityKindItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/AUnitItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/AffineConversionUnitItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/DerivedQuantityKindItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/DerivedUnitItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/DimensionFactorItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/DimensionItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/LinearConversionUnitItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/PrefixItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/PrefixedUnitItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/QuantityKindFactorItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/QuantityKindItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/SimpleQuantityKindItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/SimpleUnitItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/SystemOfQuantitiesItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/SystemOfUnitsItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/qudv/provider/UnitFactorItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/roles/provider/DisciplineItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/roles/provider/RoleManagementItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/structural/provider/GeneralRelationItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/structural/provider/RelationInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/structural/provider/StructuralElementInstanceItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/structural/provider/StructuralElementItemProvider.java
	de.dlr.sc.virsat.model.edit/src-gen/de/dlr/sc/virsat/model/dvlm/units/provider/UnitManagementItemProvider.java
	de.dlr.sc.virsat.model/src/de/dlr/sc/virsat/model/dvlm/roles/RightsHelper.java
	de.dlr.sc.virsat.model/templates-gen/org/eclipse/emf/codegen/ecore/templates/edit/ItemProvider.java
	de.dlr.sc.virsat.model/templates/edit/ItemProvider.javajet

Commit:
Missed @ Merge

Commit:
Fixing spot bugs issue - (Task #548)

- Fixed inconsistent synchronization

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Merge remote-tracking branch 'origin/development' into
feature/task_548-User_Context_Implementation

Conflicts:
	de.dlr.sc.virsat.build.test/src/de/dlr/sc/virsat/build/validator/VirSatValidatorBuilderTest.java

Commit:
Extended test case for transactional editing domain - (Task #548)

- Added test cases for verifying User Context Override.

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Merge remote-tracking branch 'origin/development' into feature/task_548-User_Context_Implementation

Commit:
Enhanced executeNoUndo in Command Stack - (Task #548)

- Enhanced CommandStack to execute a command with a given user context.
- Implemented functionality to hand over UserContext to workspace-locked
execution.
- Also added functionality to explicitly execute the builders.
- Added test cases for new functionality.
---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Deactivating automated workspace builder in server - (Task #548)

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Merge remote-tracking branch 'origin/development' into
feature/task_548-User_Context_Implementation

Conflicts:
	de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/test/AProjectTestCase.java

Commit:
Merge remote-tracking branch 'origin/development' into
feature/task_548-User_Context_Implementation

Conflicts:
	de.dlr.sc.virsat.model.calculation/src/de/dlr/sc/virsat/model/calculation/compute/EquationHelper.java
	de.dlr.sc.virsat.model/src/de/dlr/sc/virsat/model/dvlm/roles/RightsHelper.java
	de.dlr.sc.virsat.project/src/de/dlr/sc/virsat/project/resources/VirSatResourceSet.java

Commit:
Fixing test dependency - (Task #548)

- Added dependency to buidlers for the project test fragment. Otherwise
the tycho tests won't execute them. They purely won't know them.

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Fixing test dependency - (Task #548)

- Stepping back since it introduces a cyclic dependency.

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Improved Code Coverage - (Task #548)

- Due to code change in graphitti DiagramHelper, the coverage was
dropping below 80%. See PR #569

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Adding a test builder for Command Stack Test Cases - (Task #548)

- Added a builder which counts how often a builder is invoked. this is
needed for the special executenoundo function whcih can directly call
the project builder.

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Improving coverage - (Task #548)

- Test coverage is still failing for grafiti.

---
Task #548: Introduce capabilities to override current user for command
executions

Commit:
Merge remote-tracking branch 'origin/development' into feature/task_548-User_Context_Implementation

Commit:
Replaced some recording commands - (Task #584)

- Replaced RecordingCommands in DIagramHelperTest with executeAsCommand

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Replaced some recording commands - (Task #584)

- Minor fix to the test case

---
Task #584: Implement non UI Command for deleting Structural Element
Instances

Commit:
Merge remote-tracking branch 'origin/development' into
feature/task_548-User_Context_Implementation

Conflicts:
	de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/test/AllTests.java

SaMuellerDLR commented on de.dlr.sc.virsat.project.test/src/de/dlr/sc/virsat/project/test/AProjectTestCase.java:
```diff
@@ -266,4 +268,26 @@ protected void assertRetry(int retries, int milliSeconds, Runnable assertRunnabl
+			this.su = su;
+		}
+		
+		String userName;
```
Maybe put the variables before the constructor?

PhilMFischer replied:
OK

SaMuellerDLR reviewed (changes requested):
Just really minor changes.

SaMuellerDLR commented on de.dlr.sc.virsat.project/src/de/dlr/sc/virsat/project/editingDomain/VirSatWorkspaceCommandStack.java:
```diff
@@ -113,10 +114,23 @@ public void executeNoUndo(Command command) {
-		});
+			
+			// Now run the builder if requested
+			if ((executeBuilder) && !ResourcesPlugin.getWorkspace().isAutoBuilding()) {
```
Does it make sense to execute the builder if the command was not executeable or there was an exception?
I think this might make it harder to reproduce issues, say, if there is a problem with some command since the builder will change the state of the model.

PhilMFischer replied:
I wasn't realy sure how to deal with it, but actually thought about it. Since i don't have a strong opinion on either the one or the other, I am happy to follow.

PhilMFischer commented on de.dlr.sc.virsat.graphiti.test/src/de/dlr/sc/virsat/graphiti/diagram/ToolBehaviorProviderTest.java:
```diff
@@ -0,0 +1,35 @@
+public class ToolBehaviorProviderTest {
+
+	@Before
+	public void setUp() throws Exception {
```
Ok

Commit:
Changes due to review - (Task #548)

- removed empty setup and tearDown method from test cases
- fixed call to builder in Command Stack to only execute on successful
commands.
- Moved variable declaration in class.

---
Task #548: Introduce capabilities to override current user for command
executions

SaMuellerDLR reviewed (approved):
Approved!

JAmmermann-DLR commented on de.dlr.sc.virsat.project/src/de/dlr/sc/virsat/project/resources/VirSatResourceSet.java:
```diff
@@ -1005,6 +1000,8 @@ public void saveResource(Resource resource, boolean overrideWritePermissions) {
 	}
 
 	/**
+<<<<<<< HEAD
+=======
```
Should be removed

PhilMFischer replied:
Removed

franzTobiasDLR reviewed (approved):
I'm somehow feeling akword with the change that we now need to give a UserContext when saving via ResourcesSet... But as we usally do that via editing domain it should be fine I guess...

Commit:
Removing old git merge statement - (Task #548)

- Merge block was still in comment of method.

---
Task #548: Introduce capabilities to override current user for command
executions

PhilMFischer commented:
> 
> 
> I'm somehow feeling akword with the change that we now need to give a UserContext when saving via ResourcesSet... But as we usally do that via editing domain it should be fine I guess...

I Agree, at the moment i don't see a better option.
